File: utils/city.py
"""The file contains the city class for a game of Pandemic."""
from .color import Color

class City:
    """A class to represent a city in the game of Pandemic."""
    def __init__(self, name: str, color: Color, population: int):
        """Initialize a city with a name, color, and population."""
        if name:
            self.name = name
        if color in ["blue", "yellow", "black", "red"]:
            self.color = color
        else:
            raise ValueError(f"Invalid color for city {name}: {color}\n Valid colors are: blue, yellow, black, red.")
        if population > 100000:
            self.population = population
        else:
            raise ValueError(f"Invalid population for city {name}: {population}\n Population must be greater than 100,000.")
        self.disease_cubes = {Color('blue'): 0, Color('yellow'): 0, Color('black'): 0, Color('red'): 0}
        self.research_station = False
        self.neighbors = {}


    def __str__(self):
        """Return the name of the city."""
        return self.name
    
    def __repr__(self):
        """Return a representation of the city."""
        return f"City(name='{self.name}', color='{self.color}', population={self.population}, disease_cubes={self.disease_cubes}, research_station={self.research_station}, neighbors={self.neighbors})"
    
    @property
    def disease_cubes(self):
        """Return the disease cubes for the city."""
        return self.disease_cubes
    
    @property
    def neighbors(self):
        """Return the neighbors of the city."""
        return self.neighbors
    

    
    def add_neighbor(self, neighbor):
        """Add a neighbor to the city."""
        if neighbor.name not in self.neighbors.keys():
            self.neighbors[neighbor.name] = neighbor
            neighbor.add_neighbor(self)

    # def infect(self, color: Color, )

File: cpp/Private/utils/city.cpp
#include <string>
#include <vector>

namespace pandemic {
	class City {
	public:
		City(string name, Color color, int population, float position[2]) {
			this->name = name;
			this->color = color;
			this->population = population;
			this->research_station = false;
			this->outbroken = false;
			this->quarantined = false;
			this->position = position;
		}

		// getters and setters
		string getName{ return name; }
		Color getColor{ return color; }
		bool hasResearchStation{ return research_station; }
		vector<City> getNeighbors{ return neighbors; }
		bool isOutbroken{ return outbroken; }
		bool isQuarantined{ return quarantined; }
		float getPosition{ return position; }

		// methods
		void addNeighbor(City other) { neighbors.push_back(other); }

		void addDiseaseCubes(Color color, int num) {
			if (!quarantined) {
				int key = color.key;
				int curr = disease_cubes[key];

				if (num <= 0) {
					throw invalid_argument("Invalid number of cubes to add. Must add at least 1 cube");
				}
				if (num > 24) {
					throw invalid_argument("Too many cubes to add to " + name);
				}

				// too many cubes placed on city causes an outbreak
				if (curr + num > 3) {
					if (!outbroken) {
						disease_cubes[key] = 3;
						outbroken = true;
						outbreak();
					}
				}

				// normal cube add
				disease_cubes[key] += color.takeCubes(num);
			}
		}

		void removeDiseaseCubes(Color color, int num) {
			int key = color.key;
			int curr = disease_cubes[key];

			if (num <= 0) {
				throw invalid_argument("Invalid number of cubes to remove. Must remove at least 1 cube");
			}
			if (num > curr) {
				throw invalid_argument("Not enough cubes to remove from " + name);
			}

			disease_cubes[key] -= color.returnCubes(num);
		}

		void placeResearchStation() { research_station = true; }

	private:
		string name;
		Color color;
		int population;
		bool research_station;
		vector<City> neighbors;
		bool outbroken;
		bool quarantined;
		float position[2];
		vector<int> disease_cubes; // The number of disease cubes for each color, stored by index from Color Object

		void outbreak() {
			for (City neighbor : neighbors) {
				neighbor.addDiseaseCubes(color, 1);
			}
		}
	};
}